SecureCRT module updated to add automatic terminal reading support for all application versions pastt, present, and future #6702

nromey opened the pull request:
Updated securecrt.py to reflect the new window classes for Secure CRT's new versions. I'm not sure if I did this pull request thing correct and I did not compile this on my main computer, but if I've formatted it correctly i.e. the actual conditional, it will work coorrectly as this is the logic I use in the JAWS versions of the script.  This will bring support up from a 2013 scrt version to the current 2016 version.
Fixes #6302.

nromey committed:
Add files via upload

nromey committed:
Add files via upload

nromey committed:
removed root file from folder.

nromey committed:
securecrt support for all versions via regular expressioins

This is now in the right place and should be able to be merged. I ran it and it seems to work ok.

nromey commented:
Tested the change with both compiled and source. Works with all application versions and should work with upcoming releases.

nromey commented:
# #

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -4,18 +4,19 @@
 
 	def chooseNVDAObjectOverlayClasses(self, obj, clsList):
-		if obj.windowClassName in ("AfxFrameOrView80u", "AfxFrameOrView90u", "AfxFrameOrView100u") and obj.IAccessibleRole == oleacc.ROLE_SYSTEM_CLIENT:
+		if re.match("AfxFrameOrView\d{2,}u", obj.windowClassName) and obj.IAccessibleRole == oleacc.ROLE_SYSTEM_CLIENT:
```
Thanks for your work!

It'd be better to compile the regexp once as a constant and reuse it, rather than compiling every time we need to match. You can do this using the `re.compile` function and then call the `.match` method on the compiled object.

nromey commented:
I'll fix it later today, test it, and recommit it.

NER

nromey committed:
added compilaton of regexp.

I didn't think that this logic ran as much as it did. Works perfectly now.

nromey commented:
Re.compile now works. I've been running it now for a week with no adverse affects. Thanks for your tips and for not just fixing my code. I learn better with the tips.

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -2,11 +2,10 @@
 #This file is covered by the GNU General Public License.
 #See the file COPYING for more details.
-#Copyright (C) 2010-2012 NV Access Limited
+#Copyright (C) 2010-2016 NV Access Limited
```
Thanks for updating this. I'd suggest putting your own name in here too. :)

nromey replied:
Added my name in the comments section and not in documentation. Let me know if this is ok.  I didn't add it at first because I didn't know if it was proper. dded it and then removed it before last commit :D. It's back now.

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -2,11 +2,10 @@
 """App module for automatic reading of the terminal in all versions of SecureCRT
-"""
-
+Uses regular expressions to support the ever-changing window title for different incremental SecureCRT Versions."""
```
While this might be good as a comment, it's an implementation detail and I don't think it needs to be part of the module docstring.

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -2,11 +2,10 @@
-"""
-
+Uses regular expressions to support the ever-changing window title for different incremental SecureCRT Versions."""
 import oleacc
```
Blank line between docstring and imports please.

nromey replied:
I think I fixed the blank line issue. Thanks for the tip on styling.

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -16,7 +15,10 @@
 	def chooseNVDAObjectOverlayClasses(self, obj, clsList):
-		if re.match("AfxFrameOrView\d{2,}u", obj.windowClassName) and obj.IAccessibleRole == oleacc.ROLE_SYSTEM_CLIENT:
+		#initialize a compiles regexp object for all window names following the standard SecureCrt Window naming comvention for all versions
+		MATCHER = re.compile ("AfxFrameOrView\d{2,}u")
```
Because this is inside the `__init__` method, it's still going to compile every time the method runs. Instead, put it at module level (above the class). This way, it will only compile when the module is imported.

nromey replied:
I moved the instantiation outside of the class. Sorry about that, didn't think about it being compiled over and over again. Makes sense and is now fixed.

nromey committed:
made changes requested in pull comments

nromey commented:
I"m not sure if I noted my changes correctly under your comments.  Please check my indentation I believe it's correct because it finally compiled but there may be too much indentation.  It lost blocking when I moved things around so I had to redo the blocks. I also removed the regexp documentation into comments and out of the docstring. Tested ant it runs fine. Please let me know what else needs to be changed. I like this approach as you giving me tips helps me learn more rather than you just changing the files.  As I do more commits, you will hopeflly not have to request so many changes.

nromey committed:
fixed docstring again.

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -1,23 +1,28 @@
 #A part of NonVisual Desktop Access (NVDA)
 #This file is covered by the GNU General Public License.
 #See the file COPYING for more details.
-#Copyright (C) 2010-2012 NV Access Limited
```
I'd prefer the additional comments to be removed; just add your name to the copyright. Comments to describe what something does are great, but we don't need to comment when or how something was modified, as we can use the commit history for this.

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -1,23 +1,28 @@
 from NVDAObjects.window import DisplayModelEditableText, DisplayModelLiveText
 import appModuleHandler
-
+#initialize a compiles regexp object for all window names following the standard SecureCrt Window naming comvention for all versions
```
* nit: Blank line before this pleas.e
* nit: typo: compiles -> compiled

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -1,23 +1,28 @@
 import appModuleHandler
-
+#initialize a compiles regexp object for all window names following the standard SecureCrt Window naming comvention for all versions
+MATCHER = re.compile ("AfxFrameOrView\d{2,}u")
```
Perhaps rename MATCHER to something like RE_TERMINAL_WCLASS to make it clear what it's for. The reason is that this app module might also serve some additional purpose in future, at which point the name MATCHER might be confusing. Sorry; I should have picked this up in my previous review.

jcsteh commented on source/appModules/securecrt.py:
```diff
@@ -1,23 +1,28 @@
-			except ValueError:
-				pass
-			clsList[0:0] = (Terminal, DisplayModelLiveText)
+		def chooseNVDAObjectOverlayClasses(self, obj, clsList):
```
This entire function is indented one more tab than it needs to be.

jcsteh reviewed (changes requested):
Sorry; a few more changes.

nromey committed:
Add files via upload

Fixed tabs after getting a better text editor, sorry for delay. Much easier to actually hear tabs.

nromey commented:
Fixed tabs. It helps to have an editor that actually reads them.

nromey commented:
Any more issues with this? Will it be accepted?

nromey commented:
I have no idea why I closed this, but it wasn't on purpose.  Sorry about that.  Please review changes and merge if necessary or you wish it done.

nromey committed:
Update securecrt.py

Update to include proposed changes. Sorry I didn't see this until now. I thought I saw all the reviewed changes, but apparently not. All fixed now.

nromey committed:
I see what you're getting at, and I agree.  Sorry I didn't catch it either. I decided to go for Re_TerminalWinClass, not Re_TerminalWClass.  I hope that is ok. I like the Win better. If not I can put it back.

nromey commented:
I continue to test this with successive versions of scrt and it keeps working. It has problems tracking if there is lag / latencybut I don't think we can fix it, I tried adding delays if it didn't see a change after a keypress and it didn't help.  I think this is as good as we're going to get.  Work is paying for me to go to CSUN, maye I'll get to meet you. I know I've only done this little commit, but I'm trying :), and I hope I can help with other stuff too.  I really appreciate you telling me to fix things and not just doing it. Explaining the why helps me understand why iit needs to be the way it is, thanks.  Thank.

nromey commented:
By the way I made the change you requested, some day I'll get this thing right, people keep asking me when Secure CRT will work with NVDA, and I keep telling them soon. Sometimes it's easier for me to give them the compiled securecrt python file since we know the code works, we're just working on minor stuff.  Thanks again.

derekriemer commented:
Hi: In terms of delay not being enough, I think you could use [Console Timer][1] or a similar approach. in fact, if NVDA sees it as a terminal, just have people install it.

[1]: https://github.com/derekriemer/consoleTimer

nromey commented:
Interesting idea Derek.  I'll have to look into this for sure and play with it.  What do you mean have them install it.  Problem with this application is that NVDA doesn't see it as a terminal until I tell it that it is, see the regex due to changing titlebar.  Could I borrow some code from your console timer addon to taoek care of this? I think that I could intelligently figure out if lag exists and then adjust parameters from console timer on the fly as lag increases and decreases.  Being new at Python and NVDA addon and/or system programming, I'm not sure if it's even advisable or even friendly to borrow snippits from other people's addons to make parts of the say, for this thing, SecureCRT, more usable?  On ssh sites with high ping times, doing edits is somewhat impossible because when you arrow, NVDA doesn't see a change and then therefore doesn't tell you where you are when you edit.  I'll download your addon and try it on top of my modifications and then see if I could add portions of your logic into this SecureCRT.  That might require another pull request.  Good thought though, and one I'm going to look into for sure.  Thanks for your comment.

derekriemer commented:
The addon detects (Or should detect) that it was made a terminal by your
appModule.
Also, yes, you can take this code. That's fine with code marked gpl.
P.S. I tried intelligently guessing this, all my various attempts failed,
because say if you press left arrow, NVDA will wait n milliseconds then
fail to detect a cursor move and just give up. If you try to be clever, you
may end up waiting longer and longer each cursor move, which will
eventually slow you down a lot.

On Fri, Feb 3, 2017 at 4:37 PM, nromey <notifications@github.com> wrote:

> Interesting idea Derek. I'll have to look into this for sure and play with
> it. What do you mean have them install it. Problem with this application is
> that NVDA doesn't see it as a terminal until I tell it that it is, see the
> regex due to changing titlebar. Could I borrow some code from your console
> timer addon to taoek care of this? I think that I could intelligently
> figure out if lag exists and then adjust parameters from console timer on
> the fly as lag increases and decreases. Being new at Python and NVDA addon
> and/or system programming, I'm not sure if it's even advisable or even
> friendly to borrow snippits from other people's addons to make parts of the
> say, for this thing, SecureCRT, more usable? On ssh sites with high ping
> times, doing edits is somewhat impossible because when you arrow, NVDA
> doesn't see a change and then therefore doesn't tell you where you are when
> you edit. I'll download your addon and try it on top of my modifications
> and then see if I could add portions of your logic into this SecureCRT.
> That might require another pull request. Good thought though, and one I'm
> going to look into for sure. Thanks for your comment.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/6303#issuecomment-277392576>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFGivb--rYzNSGVjwVsFsul52HYlrNi1ks5rY7pFgaJpZM4Jsp5P>
> .
>



-- 

Derek Riemer: Improving the world one byte at a time!

   - University of Colorado Boulder Department of computer science, 4th
   year undergraduate student.
   - Accessibility enthusiast.
   - Proud user of the NVDA screen reader.
   - Open source enthusiast.
   - Skier.

Personal website <http://derekriemer.com>

nromey commented:
Thanks Jamie.  Here's my plan, let me know if you guys agree with my implementation.  I plan to take parts of Derek's ConsoleTimer addon and fold it into SecureCRT's module.  I also want to have a look at how Putty handles things as if this works with SecureCRT, it might be worth it to include this functionality.  I'm not as familiar with NVDA's project layout as a whole i.e. making sure that I'm programming in the correct form, so I might need your guidance in the guise of reviews, but that's the way I will learn.  As per Putty, I have had reports that tracking issues due to lag happen there.

We could have users install this addon for terminals, but if a 50 milisecond delay helps terminal racking and a hotkey to increase or decrease  cursor detection works, I'm inclined to include parts of the functionality.

Would you recommend a new pull request or can I just, provided this works the way I want it to, add functionality to my branch and see if it works, then we can decide whether or not to merge it into the main project.  Thanks again for the support guys.

jcsteh commented:
> I plan to take parts of Derek's ConsoleTimer addon and fold it into SecureCRT's module.  I also want to have a look at how Putty handles things as if this works with SecureCRT, it might be worth it to include this functionality.

This isn't something I want to accept into core. I understand the reason you want this, but if we do it this way, we'll end up with the same functionality duplicated in several places. For now, I think the add-on will suffice.

jcsteh commented:
This pull request has been open for quite a while now and there are still some outstanding (if minor) review comments that aren't addressed. Also, I'm pretty sure the latest version of the code here won't work at all because the name of the constant you've defined isn't the name you use in the method. Rather than delaying this even further and introducing yet more commits, I've created a new PR (#6847) and made the changes myself. Thanks for your work!
